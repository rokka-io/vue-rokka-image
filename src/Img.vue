<template>
  <img
    :[srcAttribute]="rokkaSrc"
    :[srcsetAttribute]="rokkaSrcset"
    :[srcAdditionalAttribute]="srcAdditionalComputed"
    :alt="alt"
    :title="title"
    :class="'rokka--attr-' + srcAttribute"
  />
</template>

<script>
import {
  generalProps,
  rokkaUrl,
  mergeDeep,
  mergeArraysDeep,
  srcset,
} from './helpers'

export default {
  props: {
    ...generalProps,
    srcAdditional: {
      type: String,
      default: null,
    },
  },

  computed: {
    // returns srcAdditional if set, otherwise the link to the rokka URL
    // Is useful, when you want to set "src" to a loading image, which then later gets replaced
    // by a lazy loader and data-src(set)
    srcAdditionalComputed() {
      return this.srcAdditionalAttribute
        ? this.srcAdditional
          ? this.srcAdditional
          : this.rokkaSrc
        : null
    },
    rokkaSrcset() {
      return srcset(this)
    },
    rokkaSrc() {
      const currentOperations = this.operations
      const currentOptions = this.options
      const currentVariables = this.variables

      const parrentOperations =
        this.$parent && this.$parent.$props && this.$parent.$props.operations
      const parrentOptions =
        this.$parent && this.$parent.$props && this.$parent.$props.options
      const parrentVariables =
        this.$parent && this.$parent.$props && this.$parent.$props.variables

      // get the current props
      // depending if passed a obj or an array
      let operations =
        Array.isArray(currentOperations) && Array.isArray(currentOperations[0])
          ? currentOperations[0] || {}
          : currentOperations
      let options = Array.isArray(currentOptions)
        ? currentOptions[0] || {}
        : currentOptions
      let variables = Array.isArray(currentVariables)
        ? currentVariables[0] || {}
        : currentVariables

      let currentProps = this.$props
      if (this.$parent.$data.isRokkaPictureTag) {
        // get the parent props
        // depending if passed a obj or an array
        const pOperations =
          Array.isArray(parrentOperations) &&
          Array.isArray(parrentOperations[0])
            ? parrentOperations[0]
            : parrentOperations
        const pOptions = Array.isArray(parrentOptions)
          ? parrentOptions[0]
          : parrentOptions
        const pVariables =
          Array.isArray(parrentVariables) && parrentVariables.length > 0
            ? parrentVariables[0]
            : parrentVariables

        operations = mergeArraysDeep(pOperations, operations)
        variables = mergeDeep(pVariables, variables)
        options = mergeDeep(pOptions, options)
        //we have the default props already from the parent in this case, so just use the added ones
        currentProps = this.$options.propsData
      }

      const url = rokkaUrl({
        ...this.$parent.$props,
        ...currentProps,
        operations,
        variables,
        options,
      })

      return url
    },
  },
}
</script>
