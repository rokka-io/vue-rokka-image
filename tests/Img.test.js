import { mount } from '@vue/test-utils'
import Img from '../src/Img'

// Now mount the component and you have the wrapper
let wrapper = mount(Img, {
  propsData: {
    organization: 'testorg',
    hash: '1234',
  },
})

describe('Component', () => {
  test('is a Vue instance', () => {
    const wrapper = mount(Img)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })
  it('renders the correct markup', () => {
    expect(wrapper.html()).toContain(
      '<img src="https://testorg.rokka.io/dynamic/1234/image.jpg" ' +
        'class="rokka--attr-src">'
    )
  })

  it('set postfix/srcset', () => {
    wrapper.setProps({
      postfix: ['1x', '2x'],
      options: [{ af: 1 }, { af: 1, dpr: 2 }],
    })
    expect(wrapper.html()).toContain(
      '<img src="https://testorg.rokka.io/dynamic/o-af-1/1234/image.jpg" ' +
        'class="rokka--attr-src" ' +
        'srcset="https://testorg.rokka.io/dynamic/o-af-1/1234/image.jpg 1x, ' +
        'https://testorg.rokka.io/dynamic/o-af-1-dpr-2/1234/image.jpg 2x"' +
        '>'
    )
  })

  it('set postfix/srcset/operations/variables', () => {
    wrapper.setProps({
      postfix: ['1x', '2x'],
      options: [{ af: 1 }, { af: 1, dpr: 2 }],
      operations: [
        {
          name: 'resize',
          options: { width: 100 },
          expressions: { height: '$height' },
        },
        {
          name: 'crop',
          options: { width: 200 },
          expressions: { height: '$height * 2' },
        },
      ],
      variables: { height: 200 },
    })
    expect(wrapper.html()).toContain(
      '<img ' +
        'src="https://testorg.rokka.io/dynamic/resize-width-100-height-%5B%24height%5D--crop-width-200-height-%5B%24height%20*%202%5D/v-height-200/o-af-1/1234/image.jpg" ' +
        'class="rokka--attr-src" ' +
        'srcset="https://testorg.rokka.io/dynamic/resize-width-100-height-%5B%24height%5D--crop-width-200-height-%5B%24height%20*%202%5D/v-height-200/o-af-1/1234/image.jpg 1x, ' +
        'https://testorg.rokka.io/dynamic/resize-width-100-height-%5B%24height%5D--crop-width-200-height-%5B%24height%20*%202%5D/v-height-200/o-af-1-dpr-2/1234/image.jpg 2x"' +
        '>'
    )
  })
})

describe('Component srcAttribute', () => {
  const wrapperSrcAttribute = mount(Img, {
    propsData: {
      organization: 'testorg',
      hash: '1234',
    },
  })
  it('srcAttribute', () => {
    wrapperSrcAttribute.setProps({
      srcAttribute: 'data-src',
      stack: 'teststack',
    })
    expect(wrapperSrcAttribute.isEmpty()).toBe(true)
    expect(wrapperSrcAttribute.is('img')).toBe(true)
    expect(wrapperSrcAttribute.attributes('class')).toBe('rokka--attr-data-src')
    expect(wrapperSrcAttribute.attributes('data-src')).toBe(
      'https://testorg.rokka.io/teststack/1234/image.jpg'
    )
    expect(wrapperSrcAttribute.attributes('src')).toBe(undefined)

    expect(wrapperSrcAttribute.html()).toContain(
      '<img class="rokka--attr-data-src" data-src="https://testorg.rokka.io/teststack/1234/image.jpg">'
    )
  })

  it('srcAdditionalAttribute', () => {
    wrapperSrcAttribute.setProps({
      srcAttribute: 'data-src',
      srcAdditionalAttribute: 'src',
    })
    expect(wrapperSrcAttribute.isEmpty()).toBe(true)
    expect(wrapperSrcAttribute.is('img')).toBe(true)
    expect(wrapperSrcAttribute.attributes('class')).toBe('rokka--attr-data-src')
    expect(wrapperSrcAttribute.attributes('data-src')).toBe(
      'https://testorg.rokka.io/teststack/1234/image.jpg'
    )
    expect(wrapperSrcAttribute.attributes('src')).toBe(
      'https://testorg.rokka.io/teststack/1234/image.jpg'
    )

    expect(wrapperSrcAttribute.html()).toContain(
      '<img class="rokka--attr-data-src" ' +
        'data-src="https://testorg.rokka.io/teststack/1234/image.jpg" ' +
        'src="https://testorg.rokka.io/teststack/1234/image.jpg">'
    )
  })

  it('srcAdditionalAttribute with srcAdditional', () => {
    wrapperSrcAttribute.setProps({
      srcAttribute: 'data-src',
      srcAdditionalAttribute: 'src',
      srcAdditional: '/loader.svg',
    })
    expect(wrapperSrcAttribute.isEmpty()).toBe(true)
    expect(wrapperSrcAttribute.is('img')).toBe(true)
    expect(wrapperSrcAttribute.attributes('class')).toBe('rokka--attr-data-src')
    expect(wrapperSrcAttribute.attributes('data-src')).toBe(
      'https://testorg.rokka.io/teststack/1234/image.jpg'
    )
    expect(wrapperSrcAttribute.attributes('src')).toBe('/loader.svg')

    expect(wrapperSrcAttribute.html()).toContain(
      '<img class="rokka--attr-data-src" ' +
        'data-src="https://testorg.rokka.io/teststack/1234/image.jpg" ' +
        'src="/loader.svg">'
    )
  })
})
