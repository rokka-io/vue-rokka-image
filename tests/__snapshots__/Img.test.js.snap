// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component renders correctly 1`] = `
<img
  class="rokka--attr-src"
  src="https://testorg.rokka.io/dynamic/1234/image.jpg"
/>
`;
